cmd/bad-key-revoker/main.go

```diff
@@ -0,0 +1,360 @@
+
+The following currently unexpired certificates that you've issued contain this public key and have been revoked:
+{{range . -}}
+{{.}}
+{{end}}`))
+
+var emailSubject = "Certificates you've issued have been revoked due to key compromise"
+
+var maxSerials = 100
+
+func (bkr *badKeyRevoker) sendMessages(mapping map[string][]string) error {
+	err := bkr.mailer.Connect()
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = bkr.mailer.Close()
+	}()
+	for email, serials := range mapping {
+		if len(serials) > maxSerials {
+			more := len(serials) - maxSerials
+			serials = serials[:maxSerials]
+			serials = append(serials, fmt.Sprintf("and %d more certificates.", more))
+		}
+		message := bytes.NewBuffer(nil)
+		err := emailTemplate.Execute(message, serials)
+		if err != nil {
+			return err
+		}
+		err = bkr.mailer.SendMail([]string{email}, emailSubject, message.String())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var keyCompromiseCode = int64(revocation.KeyCompromise)
+
+func (bkr *badKeyRevoker) invoke() (bool, error) {
+	// find rows to work on
+	uncheckedRows, err := bkr.selectUncheckedRows()
+	if err != nil {
+		return false, err
+	}
+	if len(uncheckedRows) == 0 {
+		return true, nil
+	}
+	for _, row := range uncheckedRows {
+		// find everything in keyHashToSerial with certificateStatus.status != revoked
+		unrevokedCerts, err := bkr.findUnrevoked(row)
+		if err != nil {
+			return false, err
+		}
+		if len(unrevokedCerts) == 0 {
+			// mark row as checked
+			err = bkr.markRowChecked(row)
+			if err != nil {
+				return false, err
+			}
+			continue
+		}
+		// revoke all rows, record who issued the cert
+		ownedBy := map[int64][]string{}
+		for _, cert := range unrevokedCerts {
+			_, err = bkr.raClient.AdministrativelyRevokeCertificate(context.Background(), &rapb.AdministrativelyRevokeCertificateRequest{
+				Cert:      cert.DER,
+				Code:      &keyCompromiseCode,
+				AdminName: &revokerName,
+			})
+			if err != nil {
+				return false, err
+			}
+			ownedBy[cert.RegistrationID] = append(ownedBy[cert.RegistrationID], cert.Serial)
+		}
+		// collect IDs
+		var ids []int64
+		for id := range ownedBy {
+			ids = append(ids, id)
+		}
+		// get contacts for IDs
+		idToEmail, err := bkr.resolveContacts(ids)
+		if err != nil {
+			return false, err
+		}
+		// merge any accounts with the same email
+		emailToSerials := map[string][]string{}
+		for id, email := range idToEmail {
+			emailToSerials[email] = append(emailToSerials[email], ownedBy[id]...)
+		}
+		// don't send emails to the person who revoked the certificate
+		if email, ok := idToEmail[row.RevokedBy]; ok {
+			delete(emailToSerials, email)
+		}
+		err = bkr.sendMessages(emailToSerials)
+		if err != nil {
+			return false, err
+		}
+		// mark row as checked
+		err = bkr.markRowChecked(row)
```

Comment:
I think the blockedKey should be marked as processed after the revocations are successful. That matches the rest of the logic. If the revocations were successful, but then there was an error during email sending, on the next go-through, there would be no results from `findUnrevoked`. So, might as well make that clear by putting the call to `markRowChecked` in the place where it practically takes effect.

Reply:
I'm not entirely sure I understand this comment, this appears to be what happens now? If there are no results from `findUnrevoked` it'll also get marked checked.

Reply:
You can disregard this comment. Now that the "revoke cert" and "send email" functions are moved into their own function and reordered a bit it's not relevant anymore.